situp-core/src/integrationTest/java/com/amazon/situp/integration/EndToEndIntegrationTest.java

```diff
@@ -167,38 +232,70 @@ public static ExportTraceServiceRequest getExportTraceServiceRequest(List<Resour
         final Map<String, Object> esDocSource = new HashMap<>();
         esDocSource.put("traceId", Base64.encodeBase64String(span.getTraceId().toByteArray()));
         esDocSource.put("spanId", Base64.encodeBase64String(span.getSpanId().toByteArray()));
-        esDocSource.put("parentSpanId", Base64.encodeBase64String(span.getParentSpanId().toByteArray()));
+        if (!span.getParentSpanId().isEmpty()) {
+            esDocSource.put("parentSpanId", Base64.encodeBase64String(span.getParentSpanId().toByteArray()));
+        }
         esDocSource.put("name", span.getName());
         esDocSource.put("kind", span.getKind().name());
         esDocSource.put("resource.attributes.service.name", serviceName);
         return esDocSource;
     }
 
+    @SuppressWarnings("unchecked")
+    private List<Map<String, Object>> getTargets(final List<Map<String, Object>> sources) {
+        final List<Map<String, Object>> targets = new ArrayList<>();
+        for (final Map<String, Object> source: sources) {
+            if (source.getOrDefault("target", null) != null) {
+                targets.add((Map<String, Object>) source.get("target"));
+            }
+        }
+
+        return targets;
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<Map<String, Object>> getEdges(final List<Map<String, Object>> sources) {
+        final List<Map<String, Object>> edges = new ArrayList<>();
+        for (final Map<String, Object> source: sources) {
+            if (source.getOrDefault("destination", null) != null) {
+                final Map<String, Object> edge = new HashMap<>((Map<String, Object>) source.get("destination"));
+                edge.put("serviceName", source.getOrDefault("serviceName", null));
+                edge.put("traceGroupName", source.getOrDefault("traceGroupName", null));
+                edges.add(edge);
+            }
+        }
+        return edges;
+    }
+
     private byte[] getRandomBytes(int len) {
         final byte[] bytes = new byte[len];
         RANDOM.nextBytes(bytes);
         return bytes;
     }
 
     private List<ResourceSpans> getRandomResourceSpans(int len) throws UnsupportedEncodingException {
+        spanIds = new ArrayList<>();
+        traceId = null;
         final ArrayList<ResourceSpans> spansList = new ArrayList<>();
         for(int i=0; i<len; i++) {
+            final byte[] parentId = getParentId();
+            final byte[] spanId = getSpanId();
             spansList.add(
                     getResourceSpans(
-                            UUID.randomUUID().toString(),
-                            UUID.randomUUID().toString(),
-                            getRandomBytes(8),
-                            getRandomBytes(8),
-                            getRandomBytes(16),
-                            SPAN_KINDS.get(RANDOM.nextInt(SPAN_KINDS.size()))
+                            serviceNames.get(i % serviceNames.size()),
+                            traceGroups.get(i % traceGroups.size()),
```

Comment:
Nit: Why move from random indexing to round robin indexing?

Reply:
The reason I prefer randomly generating the data and inferring which edges we should see if that its gives us much better actual coverage over time. This type of generation only tells us that the service map gives us a more narrow set of coverage.

Reply:
> The reason I prefer randomly generating the data and inferring which edges we should see if that its gives us much better actual coverage over time.

Agreed. The difficulty is how could we make assertions on edges while they are random in number and also in the "vertices"?

Reply:
Well you'd need to infer based on the data that you generate. If this works, it's not a blocker for me for now though.

Reply:
Taking another dive.